crates/wdl-engine/src/eval/v1/workflow.rs

```diff
@@ -1230,97 +1293,260 @@ impl WorkflowEvaluator {
+                            state.clone(),
+                            scope,
+                            state.subgraphs[&clause_node].clone(),
+                            max_concurrency,
+                            id,
+                        )
+                        .await?;
+
+                        // Promote all values in the scope to the parent scope
+                        let mut scopes = state.scopes.write().await;
+
+                        // Collect all names from all clauses
+                        let mut all_names: HashMap<String, Type> = HashMap::new();
+                        for clause in stmt.clauses() {
+                            if let Some(braced_scope_span) = clause.braced_scope_span() {
+                                let clause_scope = state
+                                    .document
+                                    .find_scope_by_position(braced_scope_span.start())
+                                    .expect("should have scope");
+
+                                for (name, n) in clause_scope.names() {
+                                    all_names.insert(name.to_string(), n.ty().clone());
+                                }
+                            }
+                        }
+
+                        let (parent, child) = scopes.parent_mut(scope);
+
+                        // First, promote all values that were defined in the taken branch
+                        let mut defined_names: HashSet<String> = HashSet::new();
+                        for (name, value) in child.local() {
+                            parent.insert(name.to_string(), value.clone());
+                            defined_names.insert(name.to_string());
+                        }
+
+                        // Then, set any names from other branches to `None`
+                        for (name, ty) in all_names {
+                            if !defined_names.contains(&name) {
+                                if let Type::Call(call_ty) = ty {
+                                    parent.insert(
+                                        name,
+                                        CallValue::new_unchecked(
+                                            call_ty.optional(),
+                                            Outputs::from_iter(call_ty.outputs().iter().map(
+                                                |(n, o)| {
+                                                    (n.clone(), Value::new_none(o.ty().optional()))
+                                                },
+                                            ))
+                                            .into(),
+                                        ),
+                                    );
+                                } else {
+                                    parent.insert(name, Value::new_none(ty.optional()));
+                                }
+                            }
+                        }
```

Comment:
Instead of looping twice and constructing a `HashSet`, could we just loop on the names and types from the resolved "scope union" and check to see if the name is present in the evaluation scope? If so, use the value, otherwise use a `None` value?